benchmark: add online workloads and Add Index interaction test (#1595)
pingcap/docs: dev/benchmark/add-index-with-load.md

---
title: Interaction Test on Online Workloads and `ADD INDEX` Operations
summary: This document tests the interaction effects between online workloads and `ADD INDEX` operations.
category: benchmark
---

# Interaction Test on Online Workloads and `ADD INDEX` Operations

## Test purpose

This document tests the interaction effects between online workloads and `ADD INDEX` operations in the OLTP scenario.

## Test version, time, and place

TiDB version: v3.0.1
Time: July, 2019

> Place: Beijing

Reviewer: Please add a blank line between these three lines. Otherwise, they'll come into one line at the website.

## Test environment

This test runs in a Kubernetes cluster deployed with 3 TiDB instances, 3 TiKV instances and 3 PD instances.

### Version information

| Component  |                  GitHash                   |
| :--- | :---------------------------------------- |
| TiDB  | `9e4e8da3c58c65123db5f26409759fe1847529f8` |
| TiKV  | `4151dc8878985df191b47851d67ca21365396133` |
|  PD   | `811ce0b9a1335d1b2a049fd97ef9e186f1c9efc1` |

Sysbench version：1.0.17

### TiDB parameter configuration

> TiDB, TiKV and PD use [TiDB Operator](https://github.com/pingcap/tidb-operator) configuration by default.

Reviewer: ```suggestion
TiDB, TiKV and PD all use the default [TiDB Operator](https://github.com/pingcap/tidb-operator) configuration.
```

> ### Cluster Topology

Reviewer: Keep the capitalization consistent.

```suggestion
### Cluster topology
```

|                 Machine IP                  |   Deployment instance   |
| :-------------------------------------- | :----------|
|                172.31.8.8                 |  Sysbench |
| 172.31.7.69, 172.31.5.152, 172.31.11.133 |      PD      |
| 172.31.4.172, 172.31.1.155, 172.31.9.210 |     TiKV     |
| 172.31.7.80, 172.31.5.163, 172.31.11.123 |     TiDB     |

### Online workloads simulation using Sysbench

> Use Sysbench to import **a table with 2000,000 data** into the Kubernetes cluster.

Reviewer: ```suggestion
Use Sysbench to import **a table with 2,000,000 rows of data** into the Kubernetes cluster.
```

Execute the following command to import data:

{{< copyable "shell-regular" >}}

```sh
sysbench oltp_common \
    --threads=16 \
    --rand-type=uniform \
    --db-driver=mysql \
    --mysql-db=sbtest \
    --mysql-host=$tidb_host \
    --mysql-port=$tidb_port \
    --mysql-user=root \
    prepare --tables=1 --table-size=2000000
```

Execute the following command to run the test:

{{< copyable "shell-regular" >}}

```sh
sysbench $testname \
    --threads=$threads \
    --time=300000 \
    --report-interval=15 \
    --rand-type=uniform \
    --rand-seed=$RANDOM \
    --db-driver=mysql \
    --mysql-db=sbtest \
    --mysql-host=$tidb_host \
    --mysql-port=$tidb_port \
    --mysql-user=root \
    run --tables=1 --table-size=2000000
```

## Test plan 1: frequently perform write operations to the target column of `ADD INDEX` statement

> 1. Start `oltp_read_write` test.

Reviewer: ```suggestion
1. Start the `oltp_read_write` test.
```

2. Perform at the same time with step 1: use `alter table sbtest1 add index c_idx(c)` to add an index.

> 3. Perform at the end of step 2: when the index is added successfully, stop `oltp_read_write` test.

Reviewer: ```suggestion
3. Perform at the end of step 2: when the index is added successfully, stop the `oltp_read_write` test.
```

#### `tidb_ddl_reorg_batch_size = 2048`

| tidb_ddl_reorg_worker_cnt |  add_index_durations(s) | sysbench TPS   | sysbench QPS |
| :------------------------ | :---------------------- | :------------- | :----------- |
| 1     | 918 |  243.3 |           4855 |
| 2     | 1160 |  209.9 |           4194 |
| 4     | 342 |  185.4 |           3707 |
| 8     | 1316 |  151.0 |           3027 |
| 16    | 795 |   30.5 |           679 |
| 32    | 1130 |   26.69 |           547 |
| 48    | 893 |   27.5 |           552 |

![add-index-load-1-b2048](/media/add-index-load-1-b2048.png)

#### `tidb_ddl_reorg_batch_size = 4096`

| tidb_ddl_reorg_worker_cnt |  add_index_durations(s) | sysbench TPS   | sysbench QPS |
| :------------------------ | :---------------------- | :------------- | :----------- |
| 1     | 3042 |  200.0 |           4001 |
| 2     | 3022 |  203.8 |           4076 |
| 4     | 858 |  195.5 |            3971 |
| 8     | 3015 |  177.1 |           3522 |
| 16    | 837 |   143.8 |           2875 |
| 32    | 942 |   114 |           2267 |
| 48    | 187 |   54.2 |           1416 |

![add-index-load-1-b4096](/media/add-index-load-1-b4096.png)

### Test conclusion

> When you perform frequent write operations (this test involves `UPDATE`, `INSERT` and `DELETE` operations) to the target column of `Add Index` statement, the default `Add Index` configuration has a significant impact on the online workload of the system. It is mainly because of the write conflicts caused by the concurrent `Add Index` operation and column update. The performance of the system is as follows:

Reviewer: ```suggestion
When you perform frequent write operations (this test involves `UPDATE`, `INSERT` and `DELETE` operations) to the target column of the `Add Index` statement, the default `Add Index` configuration has a significant impact on the online workload of the system. It is mainly because of the write conflicts caused by the concurrent `Add Index` operation and column update. The performance of the system is as follows:
```

> - As the two parameters increase, the value of `TiKV_prewrite_latch_wait_duration` increases significantly, slowing down the write speed.

Reviewer: ```suggestion
- As the value of `tidb_ddl_reorg_worker_cnt` and `tidb_ddl_reorg_batch_size` parameters increase, the value of `TiKV_prewrite_latch_wait_duration` increases significantly, slowing down the write speed.
```

- When the value of `tidb_ddl_reorg_worker_cnt` and `tidb_ddl_reorg_batch_size` are very large, you can execute the `admin show ddl` command to see multiple retry attempts of the DDL job, such as `Write conflict, txnStartTS 410327455965380624 is stale [try again later], ErrCount:38, SnapshotVersion: 410327228136030220`. In this situation, the `Add Index` operation takes a very long time to complete.

## Test plan 2: do not perform write operations to the target column of `ADD INDEX` statement (query-only)

> 1. Start `oltp_read_only` test.

Reviewer: Let's add an article to a countable noun.

```suggestion
1. Start the `oltp_read_only` test.
```

2. Perform at the same time with step 1: use `alter table sbtest1 add index c_idx(c)` to add an index.

> 3. Perform at the end of step 2: when the index is added successfully, stop `oltp_read_only` test.

Reviewer: ```suggestion
3. Perform at the end of step 2: when the index is added successfully, stop the `oltp_read_only` test.
```

#### `tidb_ddl_reorg_batch_size = 1024`

| tidb_ddl_reorg_worker_cnt |  add_index_durations(s) | sysbench TPS   | sysbench QPS |
| :------------------------ | :---------------------- | :------------- | :----------- |
| 1     | 91 |  536.8 |           8316 |
| 2     | 52 |  533.9 |           8165 |
| 4     | 40 |  522.4 |           7947 |
| 8     | 36 |  510 |           7860 |
| 16    | 33 |   485.5 |           7704 |
| 32    | 31 |   467.5 |           7516 |
| 48    | 30 |   562.1 |           7442 |

![add-index-load-2-b1024](/media/add-index-load-2-b1024.png)

#### `tidb_ddl_reorg_batch_size = 4096`

| tidb_ddl_reorg_worker_cnt |  add_index_durations(s) | sysbench TPS   | sysbench QPS |
| :------------------------ | :---------------------- | :------------- | :----------- |
| 1     | 103 |  502.2 |           7823 |
| 2     | 63 |  486.5 |           7672 |
| 4     | 52 |  467.4 |            7516 |
| 8     | 39 |  452.5 |           7302 |
| 16    | 35 |   447.2 |           7206 |
| 32    | 30 |   441.9 |           7057 |
| 48    | 30 |   440.1 |           7004 |

![add-index-load-2-b4096](/media/add-index-load-2-b4096.png)

### Test conclusion

> When you only perform query operations to the target column of `Add Index` statement, the effect of `Add Index` operation on online workloads is not obvious.

Reviewer: ```suggestion
When you only perform query operations to the target column of the `Add Index` statement, the effect of `Add Index` operations on online workloads is not obvious.
```

> ## Test plan 3: the target column of `ADD INDEX` statement is not related to the online workloads

Reviewer: ```suggestion
## Test plan 3: the target column of `ADD INDEX` statement is irrelevant to online workloads
```

> 1. Start `oltp_read_write` test.

Reviewer: ```suggestion
1. Start the `oltp_read_write` test.
```

2. Perform at the same time with step 1: use `alter table test add index pad_idx(pad)` to add an index.

> 3. Perform at the end of step 2: when the index is added successfully, stop `oltp_read_only` test.

Reviewer: ```suggestion
3. Perform at the end of step 2: when the index is added successfully, stop the `oltp_read_only` test.
```

#### `tidb_ddl_reorg_batch_size = 1024`

| tidb_ddl_reorg_worker_cnt |  add_index_durations(s) | sysbench TPS   | sysbench QPS |
| :------------------------ | :---------------------- | :------------- | :----------- |
| 1     | 94 |  352.4 |           6794 |
| 2     | 50 |  332 |           6493 |
| 4     | 45 |  330 |           6456 |
| 8     | 36 |  325.5 |           6324 |
| 16    | 32 |   312.5   |           6294 |
| 32    | 32 |   300.6 |           6017 |
| 48    | 31 |   279.5 |           5612 |

![add-index-load-3-b1024](/media/add-index-load-3-b1024.png)

#### `tidb_ddl_reorg_batch_size = 4096`

| tidb_ddl_reorg_worker_cnt |  add_index_durations(s) | sysbench TPS   | sysbench QPS |
| :------------------------ | :---------------------- | :------------- | :----------- |
| 1     | 116 |  325.5 |           6324 |
| 2     | 65 |  312.5 |           6290 |
| 4     | 50 |  300.6 |           6017 |
| 8     | 37 |  279.5 |           5612 |
| 16    | 34 |   250.4   |           5365 |
| 32    | 32 |   220.2 |           4924 |
| 48    | 33 |   214.8 |           4544 |

![add-index-load-3-b4096](/media/add-index-load-3-b4096.png)

### Test conclusion

> When the target column of `Add Index` statement is not related to online workloads, the effect of `Add Index` operations on the workload is not obvious.

Reviewer: ```suggestion
When the target column of the `Add Index` statement is irrelevant to online workloads, the effect of `Add Index` operations on the workload is not obvious.
```

## Summary

> - When you perform frequent write operations (including `INSERT`, `DELETE` and `UPDATE` operations) to the target column of `Add Index` statement, the default `ADD INDEX` configuration  causes relatively frequent write conflicts, which has a great impact on online workloads. At the same time, the `Add Index` operation takes a long time to complete due to continuous retry attempts. In this test, you can modify the product of `tidb_ddl_reorg_worker_cnt` and `tidb_ddl_reorg_batch_size` to 1/32 of the default value. For example, you can set `tidb_ddl_reorg_worker_cnt` to `4` and `tidb_ddl_reorg_batch_size` to `256` for better performance.

Reviewer: ```suggestion
- When you perform frequent write operations (including `INSERT`, `DELETE` and `UPDATE` operations) to the target column of the `Add Index` statement, the default `ADD INDEX` configuration causes relatively frequent write conflicts, which has a great impact on online workloads. At the same time, the `Add Index` operation takes a long time to complete due to continuous retry attempts. In this test, you can modify the product of `tidb_ddl_reorg_worker_cnt` and `tidb_ddl_reorg_batch_size` to 1/32 of the default value. For example, you can set `tidb_ddl_reorg_worker_cnt` to `4` and `tidb_ddl_reorg_batch_size` to `256` for better performance.
```

> - When only performing query operations to the target column of `Add Index` statement or the target column is not directly related to online workloads, you can use the default `ADD INDEX` configuration.

Reviewer: ```suggestion
- When only performing query operations to the target column of the `Add Index` statement or the target column is not directly related to online workloads, you can use the default `Add Index` configuration.
```
